lib/record_store/zone.rb

```diff
@@ -21,19 +21,72 @@ def each(&block)
 
       def find(name)
         return unless File.exists?(zone_path = "#{RecordStore.zones_path}/#{name}.yml")
-        Zone.from_yaml_definition(*YAML.load_file(zone_path).first)
+        load_yml_zone_definition(zone_path).first.last
+      end
+
+      def write(name, config:, records:, format: :file)
+        raise ArgumentError, "format must be :directory or :file" unless %i(file directory).include?(format)
+        name = name.chomp('.')
+        zone_file = "#{RecordStore.zones_path}/#{name}.yml"
+        zone = { name => { config: config.to_hash } }
+        records = records.map(&:to_hash).sort_by! {|r| [r.fetch(:fqdn), r.fetch(:type), r[:nsdname] || r[:address]]}
+
+        zone[name][:records] = records if format == :file
+        write_yml_file(zone_file, zone)
+        remove_record_files(name)
+        write_record_files(name, records) if format == :directory
       end
 
       private
 
-      def load_yaml_file(filename)
+      def write_yml_file(filename, data)
+        lines = data.deep_stringify_keys.to_yaml.lines
+        lines.shift if lines.first == "---\n"
+        File.write(filename, lines.join)
+      end
+
+      def load_yml_zone_definition(filename)
         result = {}
+        dir = File.dirname(filename)
         YAML.load_file(filename).each do |name, definition|
+          definition['records'] ||= []
+          Dir["#{dir}/#{name}/*__*.yml"].each do |record_file|
+            definition['records'] += load_yml_record_definitions(name, record_file)
+          end
           result[name] = Zone.from_yaml_definition(name, definition)
         end
         result
       end
 
+      def load_yml_record_definitions(name, record_file)
+        type, domain = File.basename(record_file, '.yml').split('__')
+        Array.wrap(YAML.load_file(record_file)).map do |record_definition|
+          record_definition.merge(fqdn: "#{domain}.#{name}", type: type)
+        end
+      end
+
+      def remove_record_files(name)
+        dir = "#{RecordStore.zones_path}/#{name}"
+        File.unlink(*Dir["#{dir}/*"])
+        Dir.unlink(dir)
+      rescue Errno::ENOENT
+      end
+
+      def write_record_files(name, records)
+        dir = "#{RecordStore.zones_path}/#{name}"
+        Dir.mkdir(dir)
+        records.group_by { |record| [record.fetch(:fqdn), record.fetch(:type)] }.each do |(fqdn, type), grouped_records|
+          grouped_records.each do |record|
+            record.delete(:fqdn)
+            record.delete(:type)
+            record.deep_stringify_keys!
+          end
+          grouped_records = grouped_records.first if grouped_records.size == 1
```

Comment:
This, combined with the `Array.wrap` call in `load_yml_record_definitions ` allows to have a simpler YAML file:
```yaml
ttl: 86400
address: 10.10.10.42
```

Rather than:
```yaml
- ttl: 86400
  address: 10.10.10.42
```

I'm no longer sure it's worth doing that though, we should probably have a single format for record files. But for example for CNAMEs where it's not legal to have multiple records, it's weird to force people to use a YAML array with a hash inside rather than a straight hash. WDYT?